### PR TITLE
cpu/stm32: model CPU lines in Kconfig + small refactoring

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR refactors slightly the STM32 CPU lines management:
- It contains a commit that tries to explain how the CPU info is extracted from the CPU model name
- It extends the CPU_LINE content computation in make, especially for STM32L1 and STM32G4
- It splits the `cpu/stm32/Kconfig` in several sub-Kconfig files that describe the families, the features and and models
- It introduces a new Kconfig.lines that describe all the existing CPU lines
- Now the hierarchy is CPU_FAM > CPU_LINE > CPU_MODEL. Some features defined at CPU model level could be moved at CPU line level

This PR will be useful for the STM32 clock rework, especially the Kconfig modelling part because some clock features are available in certain cpu lines, so easier to classify this way, instead of a per model basis.

### Testing procedure

- A green Murdock

### Issues/PRs references

Tick one item in #14975

### CODE REVIEW

File: cpu/stm32/Kconfig.models
Comment (benpicco) [inline cpu/stm32/Kconfig.models:68]: This is going to be messy when people start adding individual part numbers piecemeal.
I'd split that into `Kconfig.stm32f2`, `Kconfig.stm32f3` etc.

For sam0 I used a script to generate Kconfig entries for all parts where there was a vendor file #14782

This also makes working with external boards easier.

Comment (aabadie) [reply cpu/stm32/Kconfig.models:68]: For CPU models that is not possible since there are too many combination possible and there's no know list of all available models, at least not in the files available in the cmsis directory.
For CPU lines that's also not easily possible since some lines are providing features and others not.

But I agree that being able to generate/compute these lists automatically would be very useful.

Comment (aabadie) [reply cpu/stm32/Kconfig.models:68]: Also for external boards, it might still possible to just add
```kconfig
config CPU_MODEL_<complete model name>
    bool
    select CPU_LINE_<line>
```
in the board Kconfig file

Comment (benpicco) [reply cpu/stm32/Kconfig.models:68]: Since those are only used to select the appropriate header, I think it would be best to just name the options the same as the header then. So `CPU_MODEL_STM32F423XX`, `CPU_MODEL_STM32F401XC`, etc.

Comment (benpicco) [reply cpu/stm32/Kconfig.models:68]: Wait, that's already `CPU_LINE` - so what is `CPU_MODEL` used for then?

Comment (aabadie) [reply cpu/stm32/Kconfig.models:68]: > I think it would be best to just name the options the same as the header then

This is not so clear. The XX letters contains information about the pin count numbers and the ROM/RAM size. This information is used by the build system.

Comment (leandrolanzieri) [reply cpu/stm32/Kconfig.models:68]: > I'd split that into Kconfig.stm32f2, Kconfig.stm32f3 etc.

This should be possible anyway, right?

Comment (aabadie) [reply cpu/stm32/Kconfig.models:68]: But I agree that in Kconfig we could define the CPU line at board level. The problem I see is when features will be modeled in Kconfig, we'll still have to keep the CPU_MODEL definition in make, to determine the RAM/ROM sizes.

Comment (aabadie) [reply cpu/stm32/Kconfig.models]: I pushed a commit for grouping the definitions per families, except for the features.

File: cpu/stm32/Kconfig
Comment (aabadie) [inline cpu/stm32/Kconfig:124]: @leandrolanzieri, do you think it's ok this way or should I conditionally source the family Kconfig file (is it possible ?)

Comment (leandrolanzieri) [reply cpu/stm32/Kconfig:124]: Adding an `if` around a `source` directive in Kconfig does not do a conditional inclusion, but applies the condition as a dependency to all the symbols defined inside. In this case it would mean to make `CPU_FAM_F0` (which is defined inside `Kconfig.f0`) to depend on itself.

The files are split just for a matter of organization, it does not hurt to include all of them to the tree always. I think what you did is OK, we don't need the conditionals here.

Comment (aabadie) [reply cpu/stm32/Kconfig:124]: > In this case it would mean to make CPU_FAM_F0 (which is defined inside Kconfig.f0) to depend on itself.

Yeah, but if that would have been possible, I would have moved the definitions of the families back in this file.

> I think what you did is OK, we don't need the conditionals here.

Ok, great!

**Review by benpicco — APPROVED** [review]
Looks good to me.

**Comment by aabadie** [comment]
@leandrolanzieri it's all green here. Are you still OK with this PR ?

**Comment by leandrolanzieri** [comment]
The PR looks good, I just wanted to point out that this is re-introducing the CPU_LINE hierarchy which was discussed and decided against, because there was no clear definition for it to be defined on all CPUs. See https://github.com/RIOT-OS/RIOT/pull/14176#issuecomment-644660714 and https://github.com/RIOT-OS/RIOT/pull/14296

As CPU_LINE was in my original proposal when defining CPU symbols I like the idea, but I would not like to see these definitions drifting again from one CPU to another. If we would go with the approach on master these groups would be just CPU_COMMON_%. I'm not saying we should drop the idea of CPU_LINE, but I think we should agree on something we can apply to different CPUs and stick to it. @fjmolinas thoughts on this?

**Comment by fjmolinas** [comment]
> The PR looks good, I just wanted to point out that this is re-introducing the CPU_LINE hierarchy which was discussed and decided against, because there was no clear definition for it to be defined on all CPUs. See #14176 (comment) and #14296

I maintain the same opinion I had in the original PR, `CPU_LINE` is not clear to me. Also in this PR `cpu/kconfig` is not updated to reflect this change. I would like to see a definition of what `CPU_LINE` is and how it should be handled, must it always be defined? If its not used can it have the same value as `CPU_FAM`? Is it just for `STM32`? Why not `CPU_COMMON_STM32F030X8`?

**Comment by aabadie** [comment]
I'm not sure CPU_LINE could apply well to other vendors/architectures (where it might maybe correspond to CPU_FAM for example). In the case of STM32, it's an important intermediate level which allows to better determine which features are available and which are not.

For STM32, I also have doubts if we should declare all cpu models in Kconfig at CPU level and not rather let the board declare them.
My main concern is about providing support for all CPU models but without having a board corresponding to it. If we would stick to CPU lines, then having one board per line is reasonably possible and would still allow a rather complete support (someone could port a board with almost no required changes at CPU level).
Otherwise, we would have to create a Kconfig database of all STM32 cpu models that exist.

**Comment by leandrolanzieri** [comment]
> I'm not sure CPU_LINE could apply well to other vendors/architectures (where it might maybe correspond to CPU_FAM for example). In the case of STM32, it's an important intermediate level which allows to better determine which features are available and which are not.

Would `CPU_COMMON_%` work instead?

> For STM32, I also have doubts if we should declare all cpu models in Kconfig at CPU level and not rather let the board declare them.

IMO the best place to declare the CPU symbol is the CPU level, so it's consistent across CPUs. If not, we may end up with CPU symbols defined in CPU level, board level and board common level.

> My main concern is about providing support for all CPU models but without having a board corresponding to it. If we would stick to CPU lines, then having one board per line is reasonably possible and would still allow a rather complete support (someone could port a board with almost no required changes at CPU level).

I don't really see why adding a definition of a specific model when porting a board with a new CPU model would be such an issue, it's almost always just a matter of a few lines.

> Otherwise, we would have to create a Kconfig database of all STM32 cpu models that exist.

I can see how this may make it a bit easier for board porting, but I'm more of the thought of adding them as they are needed.

**Comment by benpicco** [comment]
> I don't really see why adding a definition of a specific model when porting a board with a new CPU model would be such an issue, it's almost always just a matter of a few lines.

Not all boards live inside the upstream RIOT tree.

**Comment by aabadie** [comment]
> Would CPU_COMMON_% work instead?

I don't think we should try to have all CPUs on the same line. Having a good shared structure (ARCH -> CORE -> FAM -> MODEL) is already OK. But for a given CPU level, we should also let the possibility to have intermediate convenience levels. CPU lines in STM32 are in this category.

**Comment by leandrolanzieri** [comment]
> > I don't really see why adding a definition of a specific model when porting a board with a new CPU model would be such an issue, it's almost always just a matter of a few lines.
> 
> Not all boards live inside the upstream RIOT tree.

True. @benpicco what would be in your opinion the best for this case? Out-of-tree board could always define the CPU symbol in the board's kconfig as suggested by @aabadie.

> But for a given CPU level, we should also let the possibility to have intermediate convenience levels. CPU lines in STM32 are in this category.

Based on that, and to answer @fjmolinas 's comment:

> I would like to see a definition of what CPU_LINE is and how it should be handled, must it always be defined? If its not used can it have the same value as CPU_FAM? Is it just for STM32?

This would be handled just as an STM32 specific classification and there will be no definition of a common `CPU_LINE` symbol, right? If we allow this freedom for extra hierarchies on different CPUs vendors it would be good to document this.

**Comment by aabadie** [comment]
> Out-of-tree board could always define the CPU symbol in the board's kconfig as suggested by @aabadie.

This is exactly why I'm proposing to declare the cpu model at board level for STM32. Even in the RIOT code base, because it will give a good porting example for people that are implementing their board outside of the RIOT code base.

> If we allow this freedom for extra hierarchies on different CPUs vendors it would be good to document this.

Where ?

**Comment by leandrolanzieri** [comment]
> > Out-of-tree board could always define the CPU symbol in the board's kconfig as suggested by @aabadie.
> 
> This is exactly why I'm proposing to declare the cpu model at board level for STM32. Even in the RIOT code base, because it will give a good porting example for people that are implementing their board outside of the RIOT code base.

I really don't think it is clean to define _CPU_ model symbols in _boards_ directories, specially in the RIOT code base. I just mentioned it as a possible workaround for out-of-tree boards, in the case the user is reluctant to change anything outside the board directory. What happens when there are boards that share the CPU model? We keep two definitions? Is it moved to (yet another) common?

> 
> > If we allow this freedom for extra hierarchies on different CPUs vendors it would be good to document this.
> 
> Where ?

I'd say at least here http://doc.riot-os.org/kconfig-in-riot.html#kconfig-cpu-boards-features

**Comment by aabadie** [comment]
> What happens when there are boards that share the CPU model? We keep two definitions? Is it moved to (yet another) common?

You are right that there's no clean solution for this case, which could soon (and there are already STM32 based boards in that case in RIOT).

> I'd say at least here http://doc.riot-os.org/kconfig-in-riot.html#kconfig-cpu-boards-features

Would you agree if I propose to add in a follow-up PR ?

**Comment by leandrolanzieri** [comment]
> > I'd say at least here http://doc.riot-os.org/kconfig-in-riot.html#kconfig-cpu-boards-features
> 
> Would you agree if I propose to add in a follow-up PR ?

Yes, no problem with that.

So, to clarify: this change suggests to use `CPU_LINE_` prefix instead of `CPU_COMMON_`, but without introducing a common symbol `CPU_LINE`, as an intermediate hierarchy that is used (so far) only by STM32. Should we agree to list `CPU_LINE_` as a suggested (and [well-known](http://doc.riot-os.org/kconfig-in-riot.html#kconfig-cpu-boards-features)) prefix in the case it fits properly a vendor but it is not mandatory to introduce that category for all CPUs? I think it makes sense to draw guidelines on symbol naming as much as we can, while keeping it flexible enough across vendors. @fjmolinas do you see problems going on this direction?

**Comment by aabadie** [comment]
@leandrolanzieri @fjmolinas do we agree that this PR can be merged ?

**Comment by fjmolinas** [comment]
> So, to clarify: this change suggests to use CPU_LINE_ prefix instead of CPU_COMMON_, but without introducing a common symbol CPU_LINE, as an intermediate hierarchy that is used (so far) only by STM32. Should we agree to list CPU_LINE_ as a suggested (and well-known) prefix in the case it fits properly a vendor but it is not mandatory to introduce that category for all CPUs? I think it makes sense to draw guidelines on symbol naming as much as we can, while keeping it flexible enough across vendors. @fjmolinas do you see problems going on this direction?

No problem as long as it documented.

**Comment by fjmolinas** [comment]
> @leandrolanzieri @fjmolinas do we agree that this PR can be merged ?

> Would you agree if I propose to add in a follow-up PR ?

If you open the follow up then yes ;)

**Comment by aabadie** [comment]
Had to rebase this PR after #15112 was merged. I'll open the documentation PR in a moment.

**Comment by aabadie** [comment]
> If you open the follow up then yes ;)

Here it is: #15116

**Comment by fjmolinas** [comment]
I'm ok with merging then :)

**Comment by leandrolanzieri** [comment]
> > If you open the follow up then yes ;)
> 
> Here it is: #15116

Thanks @aabadie. The Kconfig changes look good to me!

**Comment by aabadie** [comment]
All green!

**Comment by aabadie** [comment]
Before merging this one after looking at how things are organized in EFM32, I'd like to rework this PR. Please don't merge now!

**Comment by aabadie** [comment]
@leandrolanzieri I reorganized the Kconfig files: they are now grouped per families in a sub-directory. What do you think ? May I squash ?

File: cpu/stm32/Kconfig.models
Comment (aabadie) [reply cpu/stm32/Kconfig.models]: > there's no known list of all available models

In fact, all products of a given family are listed on the ST website where one can even download an Excel file with all references. So it's possible to extract a CSV and generate the complete list of models in a given family. I'll try to find some time to write a Python extractor/generator script, will be helpful and not be too difficult.

**Comment by leandrolanzieri** [comment]
> @leandrolanzieri I reorganized the Kconfig files: they are now grouped per families in a sub-directory. What do you think ?

I'm all in favor for this type of organization in folders :-)

Just one thing I noticed: any particular reason why the common symbol's (e.g. `CPU_MODEL`) defaults are left in the common Kconfig instead of splitting them as well though the folders?

**Comment by aabadie** [comment]
> any particular reason why the common symbol's (e.g. CPU_MODEL) defaults are left in the common Kconfig instead of splitting them as well though the folders?

I didn't think that was possible. I pushed a new commit reorganizing things in that direction. If we move to a semi-autogenerated list of CPU models, it will make things simpler.

**Comment by leandrolanzieri** [comment]
> I didn't think that was possible. I pushed a new commit reorganizing things in that direction. If we move to a semi-autogenerated list of CPU models, it will make things simpler.

Yep, when multiple defaults are given (on the same or multiple locations) the first one that is visible (i.e. when it has a condition the expression yields `y`) is the one that applies. See [Menu attributes section](https://www.kernel.org/doc/html/latest/kbuild/kconfig-language.html#menu-attributes) in Kconfig definition.

Looks good, please squash @aabadie !

**Comment by aabadie** [comment]
There's an issue with Murdock: the job succeeded but the status is still marked as `Expected`. Other PRs seems to have the same problem.

**Pull request merged**
